natalie/deque-template.js

```diff
@@ -0,0 +1,134 @@
+// Problem 2a): build a deque factory
+//-------
+
+// The factory itself:
+function makeDeque(values) {
+	var temp = {};
+	temp.discardPile = [];
+	temp.values = values.slice();
+	temp.length = makeDeque.arrLength;
+	temp.top = makeDeque.top;
+	temp.bottom = makeDeque.bottom;
+	temp.pop = makeDeque.pop;
+	temp.push = makeDeque.push;
+	temp.shift = makeDeque.shift;
+	temp.unshift = makeDeque.unshift;
+	temp.cut = makeDeque.cut;
+	temp.map = makeDeque.map;
+	temp.sort = makeDeque.sort;
+	temp.shuffle = makeDeque.shuffle;
+	temp.improvedShuffle = makeDeque.improvedShuffle;
+	temp.isInArray = makeDeque.isInArray;
+	temp.removeFromDiscard = makeDeque.removeFromDiscard;
+	return temp;
+}
+
+// The factory's instance methods:
+makeDeque.arrLength = function() { 	// can't use property named length;
+	return this.values.length;
+}
+
+makeDeque.top = function() {
+	return this.values[(this.values.length-1)];
+}
+
+makeDeque.bottom = function() {
+	return this.values[0];
+}
+
+makeDeque.pop = function() {
+	var discardPop = this.values.pop();
+	this.discardPile.push(discardPop);
+	return discardPop;
+}
+
+makeDeque.push = function(val) {
+	if (this.isInArray(val)) {
+		this.discardPile.splice(this.removeFromDiscard(val),1);
+		return this.values.push(val);
+	} else {
+		console.log("Nice try...you can't add that card.");
+	}
+}
+
+makeDeque.removeFromDiscard = function(value) {
+	for (i=0; i<this.discardPile.length; i++)
+		if (this.discardPile[i].id === value.id)
+		return i;
+}
+
+makeDeque.shift = function() {
+	var discardShift = this.values.shift();
+	this.discardPile.push(discardShift);
+	return discardShift;
+}
+
+makeDeque.unshift = function(val) {
+	if (this.isInArray(val)) {
+		this.discardPile.splice(this.removeFromDiscard(val),1);
+		return this.values.unshift(val);
+	} else {
+		console.log("Nice try...you can't add that card.");
+	}
+}
+
+makeDeque.cut = function() {
+	var valuesLastHalf = this.values.slice(Math.ceil(this.values.length/2));
+	this.values.splice(Math.ceil(this.values.length/2));
+	this.values = valuesLastHalf.concat(this.values);
+}
+
+makeDeque.map = function(convertValFn) {
+  return this.values.map(convertValFn);
+}
+
+makeDeque.sort = function(compareValsFn) {
+	this.values.sort(compareValsFn);
+}
+
+makeDeque.shuffle = function(shuffleFn){
```

Comment:
Somewhere in this function, the randomness is actually breaking down into a pattern. Your sortShuffle function is only shuffling about the first half of your deck. For example if you use deckOfNames.shuffle(sortShuffle) followed by deckOfNames.values about 6 times, you'll see the same 3 interval loop appearing in the way it shuffles. If you have time, go back over your sorting method and see if you can fix it to shuffle to entire deck. (or let me know if I'm just using your function incorrectly!)